cmd/hashers/argon2/calibrate.go

```diff
@@ -0,0 +1,260 @@
+					}
+					config.c.Memory = toKB(maxMemory)
+					break
+				}
+
+				currentDuration, err = probe(cmd, hasher, runs, verbose)
+				if err != nil {
+					return err
+				}
+				if verbose {
+					fmt.Fprintf(cmd.ErrOrStderr(), "  took %s with %s of memory\n", currentDuration, config.getMemFormat())
+				}
+
+				if currentDuration > desiredDuration {
+					if config.c.Memory <= toKB(adjustMemory) {
+						// adjusting the memory would now result in <= 0B
+						adjustMemory = adjustMemory >> 1
+					}
+					config.c.Memory -= toKB(adjustMemory)
+					break
+				}
+
+				// adjust config
+				config.c.Memory += toKB(adjustMemory)
+			}
+
+			if verbose {
+				fmt.Fprintf(cmd.ErrOrStderr(), "Decreasing memory to get under %s:\n", desiredDuration)
+			}
+			for {
+				currentDuration, err = probe(cmd, hasher, runs, verbose)
+				if err != nil {
+					return err
+				}
+				if verbose {
+					fmt.Fprintf(cmd.ErrOrStderr(), "  took %s with %s of memory\n", currentDuration, config.getMemFormat())
+				}
+
+				if currentDuration < desiredDuration {
+					break
+				}
+				if config.c.Memory <= toKB(adjustMemory) {
+					// adjusting the memory would now result in <= 0B
+					adjustMemory = adjustMemory >> 1
+				}
+				// adjust config
+				config.c.Memory -= toKB(adjustMemory)
+			}
+			if verbose {
+				_, _ = resultColor.Fprintf(cmd.ErrOrStderr(), "Settled on %s of memory.\n", config.getMemFormat())
+				fmt.Fprintf(cmd.ErrOrStderr(), "Increasing iterations to get over %s:\n", desiredDuration)
+			}
+			for {
+				currentDuration, err = probe(cmd, hasher, runs, verbose)
+				if err != nil {
+					return err
+				}
+				if verbose {
+					fmt.Fprintf(cmd.ErrOrStderr(), "  took %s with %d iterations\n", currentDuration, config.c.Iterations)
+				}
+
+				if currentDuration > desiredDuration {
+					config.c.Iterations -= 1
+					break
+				}
+				// adjust config
+				config.c.Iterations += 1
+			}
+
+			if verbose {
+				fmt.Fprintf(cmd.ErrOrStderr(), "Decreasing iterations to get under %s:\n", desiredDuration)
+			}
+			for {
+				currentDuration, err = probe(cmd, hasher, runs, verbose)
+				if err != nil {
+					return err
+				}
+				if verbose {
+					fmt.Fprintf(cmd.ErrOrStderr(), "  took %s with %d iterations\n", currentDuration, config.c.Iterations)
+				}
+
+				// break also when iterations is 1; this catches the case where 1 was only slightly under the desired time and took longer a bit longer on another run
+				if currentDuration < desiredDuration || config.c.Iterations == 1 {
+					break
+				}
+				// adjust config
+				config.c.Iterations -= 1
+			}
+			if verbose {
+				_, _ = resultColor.Fprintf(cmd.ErrOrStderr(), "Settled on %d iterations.\n\n", config.c.Iterations)
+			}
+
+			e := json.NewEncoder(cmd.OutOrStdout())
+			e.SetIndent("", "  ")
+			return e.Encode(config.c)
+		},
+	}
+
+	registerArgon2Flags(cmd.Flags())
+	cmd.Flags().BoolP(FlagVerbose, "v", false, "verbose output")
```

Comment:
If you want to test commands, there is no way to reset flags to default values. Having this function ensures that we always have a clean command in tests. I will outline the details of this design decision in the blog post.

The flags are defined in two places because I initially thought we could also add a command that just runs the hash function to e.g. use it for importing credentials at some point. But I will refactor that because it does not make sense as it is right now.